src/gs-image-tile.c

```diff
@@ -0,0 +1,230 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <glib/gi18n.h>
+#include <gtk/gtk.h>
+
+#include "gs-image-tile.h"
+#include "gs-star-widget.h"
+#include "gs-common.h"
+
+struct _GsImageTile
+{
+	GsAppTile	 parent_instance;
+
+	GsApp		*app;
+	GtkWidget	*app_name;
+	GtkWidget	*app_summary;
+	GtkWidget	*hover_app_name;
+	GtkWidget	*icon;
+	GtkWidget	*eventbox;
+	GtkWidget	*stack;
+	GtkWidget	*stars;
+	GtkWidget	*details_revealer;
+};
+
+G_DEFINE_TYPE (GsImageTile, gs_image_tile, GS_TYPE_APP_TILE)
+
+static GsApp *
+gs_image_tile_get_app (GsAppTile *tile)
+{
+	return GS_IMAGE_TILE (tile)->app;
+}
+
+static gboolean
+app_state_changed_idle (gpointer user_data)
+{
+	GsImageTile *tile = GS_IMAGE_TILE (user_data);
+	AtkObject *accessible;
+	GtkWidget *label;
+	gboolean installed;
+	g_autofree gchar *name = NULL;
+
+	accessible = gtk_widget_get_accessible (GTK_WIDGET (tile));
+
+	label = gtk_bin_get_child (GTK_BIN (tile->eventbox));
+	switch (gs_app_get_state (tile->app)) {
+	case AS_APP_STATE_INSTALLED:
+	case AS_APP_STATE_INSTALLING:
+	case AS_APP_STATE_REMOVING:
+	case AS_APP_STATE_UPDATABLE:
+	case AS_APP_STATE_UPDATABLE_LIVE:
+		installed = TRUE;
+		name = g_strdup_printf ("%s (%s)",
+					gs_app_get_name (tile->app),
+					_("Installed"));
+		/* TRANSLATORS: this is the small blue label on the tile
+		 * that tells the user the application is installed */
+		gtk_label_set_label (GTK_LABEL (label), _("Installed"));
+		break;
+	case AS_APP_STATE_AVAILABLE:
+	default:
+		installed = FALSE;
+		name = g_strdup (gs_app_get_name (tile->app));
+		break;
+	}
+
+	gtk_widget_set_visible (tile->eventbox, installed);
+
+	if (GTK_IS_ACCESSIBLE (accessible)) {
+		atk_object_set_name (accessible, name);
+		atk_object_set_description (accessible,
+					    gs_app_get_summary (tile->app));
+	}
+
+	g_object_unref (tile);
+	return G_SOURCE_REMOVE;
+}
+
+static void
+app_state_changed (GsApp *app, GParamSpec *pspec, GsImageTile *tile)
+{
+	g_idle_add (app_state_changed_idle, g_object_ref (tile));
+}
+
+static void
+gs_image_tile_set_app (GsAppTile *app_tile, GsApp *app)
+{
+	GsImageTile *tile = GS_IMAGE_TILE (app_tile);
+
+	g_return_if_fail (GS_IS_APP (app) || app == NULL);
+
+	if (tile->app)
+		g_signal_handlers_disconnect_by_func (tile->app,
+						      app_state_changed, tile);
+
+	g_set_object (&tile->app, app);
+	if (!app)
+		return;
+
+	if (gs_app_get_rating (tile->app) >= 0) {
+		gtk_widget_set_visible (tile->stars, TRUE);
+		gs_star_widget_set_rating (GS_STAR_WIDGET (tile->stars),
+					   gs_app_get_rating (tile->app));
+	} else {
+		gtk_widget_set_visible (tile->stars, FALSE);
+	}
+	gtk_stack_set_visible_child_name (GTK_STACK (tile->stack), "content");
+
+	g_signal_connect (tile->app, "notify::state",
+		 	  G_CALLBACK (app_state_changed), tile);
```

Comment:
You can make the callback signature just have a `GsAppTile` parameter and use `g_signal_connect_swapped()` above (we don't need the pspec)